lib/Buzz/Client/AbstractClient.php

```diff
@@ -1,77 +1,98 @@
-    protected $timeout = 5;
-    protected $verifyPeer = true;
-    protected $verifyHost = 2;
-    protected $proxy;
-
-    public function setIgnoreErrors($ignoreErrors)
-    {
-        $this->ignoreErrors = $ignoreErrors;
-    }
+    /**
+     * @var OptionsResolver
+     */
+    private $optionsResolver;
 
-    public function getIgnoreErrors()
-    {
-        return $this->ignoreErrors;
-    }
+    /**
+     * @var ParameterBag
+     */
+    private $options;
 
-    public function setMaxRedirects($maxRedirects)
+    public function __construct(array $options = [])
     {
-        $this->maxRedirects = $maxRedirects;
+        $this->options = new ParameterBag($options);
+        $this->validateOptions();
     }
 
-    public function getMaxRedirects()
+    protected function getOptionsResolver()
     {
-        return $this->maxRedirects;
-    }
+        if (null !== $this->optionsResolver) {
+            return $this->optionsResolver;
+        }
 
-    public function setTimeout($timeout)
-    {
-        $this->timeout = $timeout;
+        $this->optionsResolver = new OptionsResolver();
+        $this->configureOptions($this->optionsResolver);
     }
 
-    public function getTimeout()
+    /**
+     * Validate a set of options and return a new and shiny ParameterBag.
+     */
+    protected function validateOptions(array $options = []): ParameterBag
     {
-        return $this->timeout;
-    }
+        $parameterBag = $this->options->add($options);
+        try {
+            $parameters = $this->getOptionsResolver()->resolve($parameterBag->all());
+        } catch (\Throwable $e) {
+            // Wrap any errors.
+            throw new InvalidArgumentException($e->getMessage(), $e->getCode(), $e);
+        }
 
-    public function setVerifyPeer($verifyPeer)
-    {
-        $this->verifyPeer = $verifyPeer;
+        return new ParameterBag($parameters);
     }
 
-    public function getVerifyPeer()
+    protected function configureOptions(OptionsResolver $resolver)
     {
-        return $this->verifyPeer;
+        $resolver->setDefaults([
+            'follow_redirects' => false,
+            'max_redirects' => 5,
+            'timeout' => 30,
+            'verify_peer' => true,
+            'verify_host' => true,
+            'proxy' => null,
+        ]);
+
+        $resolver->setAllowedTypes('follow_redirects', 'boolean');
+        $resolver->setAllowedTypes('verify_peer', 'boolean');
+        $resolver->setAllowedTypes('verify_host', 'boolean');
+        $resolver->setAllowedTypes('max_redirects', 'integer');
+        $resolver->setAllowedTypes('timeout', 'integer');
+        $resolver->setAllowedTypes('timeout', 'float');
+        $resolver->setAllowedTypes('proxy', ['null', 'string']);
     }
 
-    public function getVerifyHost()
+    /**
+     * Replace all existing options with new.
+     */
+    public function withOptions(array $options): self
     {
-        return $this->verifyHost;
-    }
+        $clone = clone $this;
```

Comment:
The client looks like a service to me, so making it clonable looks weird to me

Reply:
Yeah, you are correct. I should remove "withOptions" and "withAddedOptions". A client should only be configured when we construct the object, right?